build: macOS package notarization

rvagg opened the pull request:
Works locally, and I can cleanly install .pkg's created by this on Catalina, but there are some things we need to sort out before it can be deployed:

* [x] `NOTARIZATION_ID` in the ci-release job to (1) signal that notarization is required after .pkg creation and (2) point to the correct Apple ID.
* [ ] ~~An `AC_PASSWORD` entry in the Keychain on the release machine(s) for the Apple ID identified by `NOTARIZATION_ID` (`xcrun altool --store-password-in-keychain-item "AC_PASSWORD" -u "<appleid>" -p '<password>'` (I'm documenting this in nodejs/secrets and I think there's another place we have release machine docs in nodejs/build).~~ _(giving up on keychain, see https://github.com/nodejs/build/pull/2199)_.
* [x] Xcode 11 on our release machines - we're going to have to bump everything to 11, even for Node 10. In the process we probably may as well ditch older release machines (we're still using 10.10 and 10.11 but are working on more modern machines). There's some risk involved in this but I think Build is mostly in agreement that it's minimal.
* [ ] ~~A patch to `gon`, the tool I'm using here to handle the messy bits of the notarization process. Unfortunately we have an unsigned executable inside our packages, `/usr/local/lib/node_modules/npm/node_modules/term-size/vendor/macos/term-size`, used by a dependency of a dependency of npm's. Apple reports this as an "issue" but still notarizes the package. `gon` treats all issues as fatal because this can mean the package isn't installable (see https://github.com/mitchellh/gon/pull/6), but in our case, my testing suggests that this isn't a problem at all, I think because it all ends up being command-line applications anyway. I have a proposal @ https://github.com/mitchellh/gon/pull/14 (which is used here, `"ignorable_path_issues"`) but we'll see how that goes.~~

Includes @gdams' hardened-runtime patch from https://github.com/nodejs/node/issues/29216#issuecomment-546932966

@nodejs/build @nodejs/platform-macos 

Ref: https://github.com/nodejs/node/issues/29216

gdams commented:
Thanks for taking this on @rvagg! Let me know if you have any issues with my patch!

gdams commented:
> Unfortunately we have an unsigned executable inside our packages, /usr/local/lib/node_modules/npm/node_modules/term-size/vendor/macos/term-size, used by a dependency of a dependency of npm's. Apple reports this as an "issue" but still notarizes the package.

The only thing to be aware of is that going forward this won't work as Apple has only temporarily loosened the notarization requirements. (That warning will soon become an error).

One possible approach would be to codesign everything in node_modules with the Foundation codesign cert. You can add `--deep` I believe to loop through all files and folders.

rvagg commented:
> One possible approach would be to codesign everything in node_modules with the Foundation codesign cert. You can add --deep I believe to loop through all files and folders.

How sure are you of this? It doesn't raise any issues on the install size. Have they stated that all "issues" are going to be fatal in the future?

That will mean that we're taking responsibility, reputationally, for [term-size](https://github.com/sindresorhus/term-size). I suppose we're doing that anyway for everything we package, so maybe there's no difference? It just doesn't _feel_ right to be signing other people's binaries!

gdams commented:
> Have they stated that all "issues" are going to be fatal in the future?

See https://developer.apple.com/news/?id=12232019a. Essentially after February 3 2020, the original requirements will come back into place (Those warnings about Unsigned code will turn to errors)

> That will mean that we're taking responsibility, reputationally, for term-size. I suppose we're doing that anyway for everything we package, so maybe there's no difference? It just doesn't feel right to be signing other people's binaries!

I'm kind of torn on this, part of me thinks that we should be taking responsibility, reputationally for everything that we ship but I'd much rather work with the developers to put it right at their end. The main issue is that it requires a codesign cert which cost money and therefore can be challenging for small open-source projects. 

If we took on the role of codesigning all the code we would also have to assume a set of entitlements. One possible solution might be to check each binary and only sign what is not already signed (term-size in this case)

rvagg commented:
OK, so I've pushed a commit that does a codesign for `term-size` (`"$PKGDIR"/lib/node_modules/npm/node_modules/term-size/vendor/macos/term-size`) as well as `node` (`"$PKGDIR"/bin/node`) and this gets it past the notarization without any warnings. It's using the same entitlements plist, which is probably overkill, but I don't think that matters too much does it @gdams? Or maybe we should craft a new one just for term-size? I don't really know how but I can see some things in there that probably shouldn't (allow-jit, allow-unsigned-executable-memory, disable-executable-page-protection, probably the other two as well!).

I asked sindresorhus about bundling a signed version of the executable but I also think it's reasonable to say no to that request. We just have to be comfortable with codesigning code we aren't really responsible for.

gdams commented:
@rvagg I would leave it as the same entitlements for now. We can always narrow it down later, if you've got a clean report without any warnings then you should be good to go.

Just be prepared for more packages to possibly appear later on that also aren't codesigned.

gdams commented on tools/osx-gon-config.json.tmpl:
```diff
@@ -0,0 +1,14 @@
+     "password": "@keychain:AC_PASSWORD"
+  },
+
+  "ignorable_path_issues": "node-v.*\\.pkg/npm-v.*\\.pkg Contents/Payload/usr/local/lib/node_modules/npm/node_modules/term-size/vendor/macos/term-size"
```
I assume that this can go now that you're codesigning it?

gdams commented on tools/osx-codesign.sh:
```diff
@@ -8,4 +8,10 @@ if [ "X$SIGN" == "X" ]; then
+  --options runtime \
+  --timestamp \
+  "$PKGDIR"/bin/node \
+  "$PKGDIR"/lib/node_modules/npm/node_modules/term-size/vendor/macos/term-size
```
maybe add a comment here linking to this issue

sam-github commented:
@rvagg I know sindrehous by reputation, I don't suspect him even a bit of being untrustworthy.... but I didn't find the code to the term-size executable, so we are notarizing an executable that has unknown capabilities. Of course, even if the code was reviewable, that doesn't guarantee the executable came from that code. Maybe that's a good thing about notarization, if its notarized as only being able to run and query term widths (whatever capability that is), but then the exe tries to do anything else, alaram bells will go off? For that purpose, its probably worth the extra effort to reduce its capabilities.

I was looking for the code because if its a feature that is important enough npm is going to depend on it, then perhaps it should be a uv/node.js API. I can see why he provides the two executables rather than a C++ addon, its a clever workaround for the nastiness of addons, but I wish there was another way.

sindresorhus commented:
> I asked sindresorhus about bundling a signed version of the executable but I also think it's reasonable to say no to that request. We just have to be comfortable with codesigning code we aren't really responsible for.

I will look into getting it notarized.

I'm also happy to have you build the binary, notarize it, and include the binary you built into `term-size` if that would be even better? It only has to be done once as the code is unlikely to change. The source is here and it's tiny: https://github.com/sindresorhus/macos-term-size/blob/master/term-size.c, so you could probably easily decompile the existing binary to confirm too.

> I was looking for the code because if its a feature that is important enough npm is going to depend on it, then perhaps it should be a uv/node.js API. I can see why he provides the two executables rather than a C++ addon, its a clever workaround for the nastiness of addons, but I wish there was another way.

I think it's definitely something that should be supported by Node.js built-in.

> I can see why he provides the two executables rather than a C++ addon, its a clever workaround for the nastiness of addons, but I wish there was another way.

If only Node.js had built-in FFI support, we could solve most of these problems in user-land without binaries or messy native addons...

rvagg commented:
@nodejs/build would there be any objections to submitting a PR with a codesigned version of this binary to @sindresorhus' term-size repo using the "Node.js Foundation" codesigning certificate?
It would set a precedent of course, not just for this repo but anything that npm wants to bundle that has an executable. But a single signing means that we'll compile it and be verifying the specific code that the binary comes from and locks it in place, gating updates via access to the certificate.

btw @sindresorhus if you wanted to do this yourself you only need to codesign, not notarize, but you'd need to compile with a hardened runtime profile too.

sam-github commented:
No objection from me.

I thought it might be possible to do what term-size does in node, but it didn't work. It should not be so hard :-(, maybe worth opening an issue about. But that's a bit tangential to getting what we have notarized now.

```
% nvm run 13 -p 'require("readline").createInterface({input: require("fs").createReadStream("/dev/tty"), output: require("fs").createWriteStream("/dev/tty")}).getCursorPos()'                                                                                                                    
Running node v13.6.0 (npm v6.13.4)                                                                                                                                            
readline.js:755                                                                                                                                                               
  const strBeforeCursor = this._prompt + this.line.substring(0, this.cursor);
                                                   ^
                                                                                                                                                                              
TypeError: Cannot read property 'substring' of undefined                                                                                                                      
    at Interface.getCursorPos (readline.js:755:52)                        
```

rvagg commented:
Removed the "ignore warnings" feature, added a note in codesign.sh and pointed back here.

This is ready for review I think. It's still pending new build machines that we can switch to and start producing these, plus some agreement from @nodejs/build that we're committing to switching to Xcode 11 and the latest macOS for all of our release builds.

I've removed this env var from the pkg build script in iojs+release in ci-release: `NOTARIZATION_ID=build@iojs.org`, the lack of that env var will mean the notarization isn't invoked. But when it's present it'll try to run notarization, so we should put that back when we're ready.

I believe that macOS release machines are pending in https://github.com/nodejs/build/issues/1695, we have some new boxes being set up at nearForm. We're currently having a bit of trouble getting responses from MacStadium about whether we can get access to newer hardware (and software). Tracking that in https://github.com/nodejs/build/issues/1732

richardlau commented:
> 
> 
> @nodejs/build would there be any objections to submitting a PR with a codesigned version of this binary to @sindresorhus' term-size repo using the "Node.js Foundation" codesigning certificate?
> It would set a precedent of course, not just for this repo but anything that npm wants to bundle that has an executable. But a single signing means that we'll compile it and be verifying the specific code that the binary comes from and locks it in place, gating updates via access to the certificate.
> 

If we're codesigning the `term-size` binary with the Node.js Foundation certificate I'd much prefer that we compiled it from [source](https://github.com/sindresorhus/macos-term-size/blob/master/term-size.c), codesigned that and then PR'ed that into https://github.com/sindresorhus/term-size. It's a much simpler audit trail than trying to track backwards that the binary hasn't been tampered with (and plus would mean signing the binary once rather than for every release we do).

sindresorhus commented:
> If we're codesigning the term-size binary with the Node.js Foundation certificate I'd much prefer that we compiled it from source , codesigned that and then PR'ed that into sindresorhus/term-size.

Yes, that's what I'm suggesting.

rvagg commented:
Yes, that's a TODO, but it needs some bandwidth to write a signing script, figuring out the right hardening configuration, submit that back to macos-term-size and then a binary to term-size.

I'm deferring that right now because I don't have that bandwidth, maybe later in the week.

If this is a requirement to merging this PR, know that it'll delay getting notarized builds of Node out because we have to go through a chain of releases of term-size -> boxen -> npm, then an npm PR back to nodejs/node, then a backport of that npm to all active release lines.

Which is fine, but expect Catalina users to squawk louder as the Apple turn on strict-mode on February 3rd and they can no longer install .pkgs.

Getting release machines sorted out may be more of a bother than we anticipate too.

rvagg commented:
btw there is a shortcut to this dependency problem of waiting for npm to land with a signed term-size, we just merge one in to our tree inside npm, just got to get it compiled and signed though.

rvagg commented:
https://github.com/nodejs-private/secrets/pull/75 has hopefully enough stuff to get a new release machine set up, @sam-github @AshCripps let me know how you go with the nearForm machines. From memory, doing things from cmdline is a pain in the backside because it requires proper keyboard interactive passwords for some key things. I don't know what the current state of the ansible scripts is, hopefully that has all that's needed. 🤞

rvagg commented:
* codesigning in the macos-term-size build script: https://github.com/sindresorhus/macos-term-size/pull/3
* signed term-size binary submitted to term-size: https://github.com/sindresorhus/term-size/pull/16
* backed out the local term-size bundling here

tested and it's producing good .pkg files that apple is happy to notarize.

MajuMadhusudanan commented:
Which build this fix should be available ?

slohiya commented:
When will this be available? Will this be available on the LTS (12) version of Node?

AndrewTrapani commented:
I didn't notice (or maybe it's not in this commit), the CI may also "staple" the notarization "ticket" to the pkg. This allows users without active internet to run the installer without issues.

1. Notarize succeeds
2. Run the stapling command `xcrun stapler staple MyPKG.pkg`

I've verified this process with our app which is distributed in a ZIP or DMG.

Presumably, that is all that is required to staple the ticket to the pkg, even if it had KEXTs in it:
https://forums.developer.apple.com/thread/116692

Describing the validation process correctly (AFAIK):

> Stapling is a post-process step that downloads the Notarization Ticket and attaches it to the binary (before distribution) so that the end user's machine can verify the Notarization without a network connection back to Apple being available at all.

https://forums.developer.apple.com/thread/116279

MajuMadhusudanan commented:
when will notarisation fix will be available for download?

rvagg commented:
@AndrewTrapani this PR should be doing that already, it's using [gon](https://github.com/mitchellh/gon) and the config in here has `staple = true`.

@MajuMadhusudanan no ETA, we're waiting on some infra changes.

MajuMadhusudanan commented:
> @AndrewTrapani this PR should be doing that already, it's using [gon](https://github.com/mitchellh/gon) and the config in here has `staple = true`.
> 
> @MajuMadhusudanan no ETA, we're waiting on some infra changes.

Our build notarization cycle is failing due to this issue. Any updates regarding the release timeline?

AshCripps commented:
@rvagg the release machine _should_ be good to go on this (I can see the test machines are in CI but I cant see CI-release) only thing left on my end is to add the build-release ssh but Ill coordinate with @sam-github tomorrow to get it on there

slohiya commented:
Any update on this would be appreciated as this issue is completely blocking us from releasing mac version of our application.

rvagg commented:
Am making some progress on a release job for this: https://github.com/nodejs/build/pull/2199 there's a lot of awkwardness in doing this on non-UI servers.

rvagg commented:
First notarized build through our CI system: https://nodejs.org/download/test/v14.0.0-test20200303e19f0942ea/node-v14.0.0-test20200303e19f0942ea.pkg (master plus these changes). I haven't tested it myself yet but would love feedback from anyone in this thread running Catalina that would like to help out by verifying.

This needs to be applied all other release lines too, I'll do that in separate PRs, expand https://github.com/nodejs/build/pull/2199 to enable it and run some test builds for each of them.

andrewhughes101 commented:
Can confirm https://nodejs.org/download/test/v14.0.0-test20200303e19f0942ea/node-v14.0.0-test20200303e19f0942ea.pkg doesn't show any warnings when downloading on Catalina

gwhenne commented:
Also can confirm it downloads and installs without warnings.

MajuMadhusudanan commented:
> First notarized build through our CI system: https://nodejs.org/download/test/v14.0.0-test20200303e19f0942ea/node-v14.0.0-test20200303e19f0942ea.pkg (master plus these changes). I haven't tested it myself yet but would love feedback from anyone in this thread running Catalina that would like to help out by verifying.
> 
> This needs to be applied all other release lines too, I'll do that in separate PRs, expand [nodejs/build#2199](https://github.com/nodejs/build/pull/2199) to enable it and run some test builds for each of them.

It installs without warnings. Now I'm bundling it into my package file and submitting our package file for notarization

AshCripps commented:
> It installs without warnings. Now I'm bundling it into my package file and submitting our package file for notarization

@MajuMadhusudanan I would advise against that, thats a notorized build of master which is not stable. Notorization should be in all supported lines very soon.

AndrewTrapani commented:
I downloaded the linked pkg and made sure to only "Save File" and not have it auto open.

Turned off my internet after download was complete

Double clicked on the pkg

No warnings.

I did everything short of clicking "Install", because I did this on my dev machine.

I'm not sure where Gatekeeper validation occurs in a PKG and so I'll try installing it on a fresh machine in a few minutes.

AndrewTrapani commented:
I copied the PKG linked above to a MacOS Catalina 10.15.2 laptop.

Disconnected it from the internet.

Installed the PKG.

Opened a new terminal tab and ran npm. `npm --version` is 6.13.7

I was able to install a package `npm install express`

I was able to run `sudo npm install npm@latest -g` and it updated to 6.14.2.

Note that terminal does not have any security exceptions on this computer. Can be found under System Preferences -> Security and Privacy -> Privacy tab -> Developer Tools and make sure Terminal is unchecked.

Under System Preferences -> Security and Privacy -> General I have the radio button "Allow apps downloaded from:" "* App store and identified developers"

As far as I can tell, code signing, notarizing, and stapling the ticket to the PKG worked.

rvagg commented on BUILDING.md:
```diff
@@ -167,7 +167,7 @@ Binaries at <https://nodejs.org/download/release/> are produced on:
 | --------------------- | ------------------------------------------------------------------------ |
 | aix-ppc64             | AIX 7.1 TL05 on PPC64BE with GCC 6                                       |
-| darwin-x64 (and .pkg) | macOS 10.11, Xcode Command Line Tools 10 with -mmacosx-version-min=10.10 |
+| darwin-x64 (and .pkg) | macOS 10.15, Xcode Command Line Tools 11 with -mmacosx-version-min=10.10 |
```
This should probably carry an additional note when backported to 13, 12 and 10 saying what it _was_ on up to a specific version, since we'll be changing mid-major.

rvagg commented:
Added a new commit to update BUILDING.md, squashed and rebased to master. These 3 commits should be backported as a package to 13, 12 and 10. Plan of attack is posted here: https://github.com/nodejs/build/pull/2199#issuecomment-594253577, first step is to get this merged into master so nightlies start coming out notarized and then a controlled backport and version sync with CI after that.

Have requested a CI run on this final version.

Please review @nodejs/build & anyone else!

sam-github commented:
I'm hoping they will move to using pure-js, the binary is unnecessary (https://github.com/sindresorhus/term-size/issues/15).

rvagg commented:
Yeah, also, this binary is super simple and not very likely to change in the future. I've made it clear that we consider this a special case and we don't want to make a habit of doing this into the future. The fact is though, that notarization is a major hassle for people who aren't developing serious macOS or iOS apps.

rvagg commented:
Really need a review or two from @nodejs/collaborators who feel sufficiently qualified before I can land this and we can move forward.

rvagg commented:
Landed in 811b3a9931b6...c933cbfd74a8

will work on getting nightlies notarized over @ https://github.com/nodejs/build/pull/2199

rvagg commented:
Marked as dont-land-on for all release lines for now, will remove each one as build infra is ready for it. I'm hoping to have it sorted for the next 13.x which is [scheduled](https://github.com/nodejs/Release/issues/487) for the 17th.

trevors20 commented:
@rvagg , could I confirm your scheduling statement above?  Looking at the schedule, it looks like this would be available to be downloaded from https://nodejs.org/ starting April 2020?
Thanks for working on this.

rvagg commented:
https://nodejs.org/download/nightly/v14.0.0-nightly20200312e45874723f/

nightlies are now using this, .pkg files from here forward in nightlies should all be notarized.

In the process I've discovered that the new `codesign` `--entitlements` flag we're using isn't suported by the older Xcode (I think it's 10!) we're using for previous versions. So that means the upgrade path here is even tighter and the `NOTARIZATION_ID` check included here isn't enough to switch between notarization and plain codesigning.

**Backporting these commits to previous versions needs to be done simultaneously with switches in compile versions** or .pkg building will fail:

 * https://github.com/nodejs/build/pull/2202 for 13.x
 * https://github.com/nodejs/build/pull/2203 for 12.x and 10.x

It's possible to merge those PRs before backporting, but included in the commits here is BUILDING.md edits that say we're using the newer build machines. So it should be synchronized both ways.

cclauss commented:
We could call [`XcodeVersion()`](https://github.com/nodejs/node-gyp/blob/master/gyp/pylib/gyp/xcode_emulation.py#L1388) to determine if we are able to honor the request to sign the output.

rvagg commented:
I've opened https://github.com/nodejs/node/pull/32243 to get this backported to 13.x. I don't have a strong opinion on whether this goes out in the next 13.x or the one after that. We'll have a week or so of nightlies to go on so I'm confident that all will be fine and any major surprises will be exposed by then.

@cclauss python doesn't get involved in this but we could do something similar with Bash if necessary. I'm just thinking that we're the only ones likely to be even calling tools/osx-codesign.sh anyway, people downstream from us building their own binaries are not likely to be using this, so it's probably not something we really need to care too deeply about. However, if someone thinks it's important then feel free to open a PR to branch on Xcode version.

rvagg commented:
https://nodejs.org/download/nightly/v14.0.0-nightly20200313edb3ffb003/node-v14.0.0-nightly20200313edb3ffb003.pkg

brand new notarized nightly, automated this time (last one was a rerun of yesterday's)

MajuMadhusudanan commented:
Is this fix available on any production build?

rvagg commented:
queued in the next 13.x, #32376

likely will be in the 12.x and 10.x after that if we don't get major problems reported on the 13.x

rvagg commented:
Hey folks in this issue, @sindresorhus has a new release of `term-size` out so there's a new binary that's not signed. I don't know if npm is still using `term-size` and what a timeline might look like for it making its way here but the flag is that we may have notarization issues if/once it bubbles its way up here.

We previously contributed a signed binary to make everything happy, we may need to do so again. Heads-up: https://github.com/sindresorhus/macos-term-size/pull/5

I'll let someone else make the call wrt necessity and urgency but I can help get the signing done if that's needed, and the help is needed.
